### ISSUE TITLE
Crash with DataParallelTable and clearState

### ISSUE THREAD

**Issue by mdelakis** [soumith/cudnn.torch#172]
When folding into a `nn.DataParallelTable` a model and calling `forward` after a `clearState` produces a crash. Example code snippet:

```Lua
require 'cutorch'
require 'cunn'
require 'cudnn'

local net  = nn.Sequential()
net:add(cudnn.SpatialConvolution(3, 128, 3, 3, 1, 1) )
net:cuda()
local dpt = nn.DataParallelTable(1)
dpt:add(net, {1, 2})
dpt:threads(function() local cudnn = require 'cudnn' end)
local input = cutorch.createCudaHostTensor(32, 3, 224, 224):zero()

dpt:forward(input)
dpt:clearState()
dpt:forward(input)
```

Produces output:
```
...
/usr/local/share/lua/5.1/cudnn/SpatialConvolution.lua:370: attempt to perform arithmetic on a nil value
...
```
It was tested with both `master` and `R5` versions with no change.

The above code does not crash:

* If using `cunn` backend
* If disabling multi-threading in `DataParallelTable`, so it may be a `cunn` issue.

**Comment by ngimel** [comment]
DataParallelTable in threaded mode is not handling clearState() correctly. I'll submit a PR, in the meantime, instead of ```dpt:clearState()``` you can use ```dpt.impl:exec(function(m) m:clearState() end)```

**Comment by mdelakis** [comment]
Thanx the workaround fixed the problem!

**Comment by mdelakis** [comment]
Fixed via torch/cunn#257
